File: php/tests/GildedRoseTest.php
<?php

declare(strict_types=1);

namespace Tests;

use GildedRose\GildedRose;
use GildedRose\Item;
use PHPUnit\Framework\TestCase;

class GildedRoseTest extends TestCase
{
    //-----------------
    // normal item
    //-----------------
    public function testUpdatesNormalItemsBeforeSellDate(): void
    {
        // arrange
        $items = [new Item('normal', 5, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(4, $items[0]->sellIn);
        $this->assertSame(9, $items[0]->quality);
    }

    public function testUpdatesNormalItemsOnSellDate(): void
    {
        // arrange
        $items = [new Item('normal', 0, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-1, $items[0]->sellIn);
        $this->assertSame(8, $items[0]->quality);
    }

    public function testUpdatesNormalItemsAfterSellDate(): void
    {
        // arrange
        $items = [new Item('normal', -5, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-6, $items[0]->sellIn);
        $this->assertSame(8, $items[0]->quality);
    }

    public function testUpdatesNormalItemsWithAQualityOf0(): void
    {
        // arrange
        $items = [new Item('normal', 5, 0)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(4, $items[0]->sellIn);
        $this->assertSame(0, $items[0]->quality);
    }

    //-----------------
    // Brie item
    //-----------------
    public function testUpdatesBrieItemsBeforeSellDate(): void
    {
        // arrange
        $items = [new Item('Aged Brie', 5, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(4, $items[0]->sellIn);
        $this->assertSame(11, $items[0]->quality);
    }

    public function testUpdatesBrieItemsBeforeSellDateWithMaximumQuality(): void
    {
        // arrange
        $items = [new Item('Aged Brie', 5, 50)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(4, $items[0]->sellIn);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testUpdatesBrieItemsOnSellDate(): void
    {
        // arrange
        $items = [new Item('Aged Brie', 0, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-1, $items[0]->sellIn);
        $this->assertSame(12, $items[0]->quality);
    }

    public function testUpdatesBrieItemsOnSellDateNearMaximumQuality(): void
    {
        // arrange
        $items = [new Item('Aged Brie', 0, 49)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-1, $items[0]->sellIn);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testUpdatesBrieItemsOnSellDateWithMaximumQuality(): void
    {
        // arrange
        $items = [new Item('Aged Brie', 0, 50)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-1, $items[0]->sellIn);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testUpdatesBrieItemsAfterSellDate(): void
    {
        // arrange
        $items = [new Item('Aged Brie', -10, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-11, $items[0]->sellIn);
        $this->assertSame(12, $items[0]->quality);
    }

    public function testUpdatesBrieItemsAfterSellDateWithMaximumQuality(): void
    {
        // arrange
        $items = [new Item('Aged Brie', -10, 50)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-11, $items[0]->sellIn);
        $this->assertSame(50, $items[0]->quality);
    }

    //-----------------
    // Sulfuras item
    //-----------------
    public function testUpdatesSulfurasItemsBeforeSellDate(): void
    {
        // arrange
        $items = [new Item('Sulfuras, Hand of Ragnaros', 5, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(5, $items[0]->sellIn);
        $this->assertSame(10, $items[0]->quality);
    }

    public function testUpdatesSulfurasItemsOnSellDate(): void
    {
        // arrange
        $items = [new Item('Sulfuras, Hand of Ragnaros', 0, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(0, $items[0]->sellIn);
        $this->assertSame(10, $items[0]->quality);
    }

    public function testUpdatesSulfurasItemsAfterSellDate(): void
    {
        // arrange
        $items = [new Item('Sulfuras, Hand of Ragnaros', -1, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-1, $items[0]->sellIn);
        $this->assertSame(10, $items[0]->quality);
    }

    //-----------------
    // Backstage Pass
    //-----------------
    public function testUpdatesBackstagePassItemsLongBeforeSellDate(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', 11, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(10, $items[0]->sellIn);
        $this->assertSame(11, $items[0]->quality);
    }

    public function testUpdatesBackstagePassItemsCloseToBeforeSellDate(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', 10, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(9, $items[0]->sellIn);
        $this->assertSame(12, $items[0]->quality);
    }

    public function testUpdatesBackstagePassItemsCloseToSellDateAtMaximumQuality(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', 10, 50)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(9, $items[0]->sellIn);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testUpdatesBackstagePassItemsVeryCloseToSellDate(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', 5, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(4, $items[0]->sellIn);
        $this->assertSame(13, $items[0]->quality);
    }

    public function testUpdatesBackstagePassItemsVeryCloseToSellDateAtMaximumQuality(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', 5, 50)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(4, $items[0]->sellIn);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testUpdatesBackstagePassItemsWithOneDayLeftToSellDateAtMaximumQuality(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', 1, 50)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(0, $items[0]->sellIn);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testUpdatesBackstagePassItemsOnSellDate(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', 0, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-1, $items[0]->sellIn);
        $this->assertSame(0, $items[0]->quality);
    }

    public function testUpdatesBackstagePassItemsAfterSellDate(): void
    {
        // arrange
        $items = [new Item('Backstage passes to a TAFKAL80ETC concert', -1, 10)];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame(-2, $items[0]->sellIn);
        $this->assertSame(0, $items[0]->quality);
    }

    /**
     * conjured item
     *
     * @dataProvider dataProviderConjuredForSellDateAndQuality
     * @param array $input
     * @param array $expected
     * @return void
     */
    public function testUpdatesConjuredItem(array $input, array $expected): void
    {
        // arrange
        $items = [new Item('Conjured', $input['sellIn'], $input['quality'])];
        $app = new GildedRose($items);

        // act
        $app->updateQuality();

        // assert
        $this->assertSame($expected['sellIn'], $items[0]->sellIn);
        $this->assertSame($expected['quality'], $items[0]->quality);
    }

    /**
     * 測試情境
     * BeforeSellDate
     * OnSellDate
     * AfterSellDate
     * WithAQualityOf0
     * CloseToMinQualityBeforeSellDate
     * CloseToMinQualityOnSellDate
     * CloseToMinQualityAfterSellDate
     *
     * @return array[]
     */
    public function dataProviderConjuredForSellDateAndQuality(): array
    {
        return [
            'BeforeSellDate' => [
                'input' => [
                    'sellIn' => 5,
                    'quality' => 10,
                ],
                'expected' => [
                    'sellIn' => 4,
                    'quality' => 8,
                ],
            ],
            'OnSellDate' => [
                'input' => [
                    'sellIn' => 0,
                    'quality' => 10,
                ],
                'expected' => [
                    'sellIn' => -1,
                    'quality' => 6,
                ],
            ],
            'AfterSellDate' => [
                'input' => [
                    'sellIn' => -5,
                    'quality' => 10,
                ],
                'expected' => [
                    'sellIn' => -6,
                    'quality' => 6,
                ],
            ],
            'WithAQualityOf0' => [
                'input' => [
                    'sellIn' => 5,
                    'quality' => 0,
                ],
                'expected' => [
                    'sellIn' => 4,
                    'quality' => 0,
                ],
            ],
            'CloseToMinQualityBeforeSellDate' => [
                'input' => [
                    'sellIn' => 5,
                    'quality' => 1,
                ],
                'expected' => [
                    'sellIn' => 4,
                    'quality' => 0,
                ],
            ],
            'CloseToMinQualityOnSellDate' => [
                'input' => [
                    'sellIn' => 0,
                    'quality' => 1,
                ],
                'expected' => [
                    'sellIn' => -1,
                    'quality' => 0,
                ],
            ],
            'CloseToMinQualityAfterSellDate' => [
                'input' => [
                    'sellIn' => -2,
                    'quality' => 1,
                ],
                'expected' => [
                    'sellIn' => -3,
                    'quality' => 0,
                ],
            ],
        ];
    }
}

File: csharpcore/GildedRoseTests/GildedRoseTest.cs
using Xunit;
using System.Collections.Generic;
using GildedRoseKata;

namespace GildedRoseTests
{
    public class GildedRoseTest
    {
        [Fact]
        public void testUpdatesNormalItemsBeforeSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "normal", SellIn = 5, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(4, Items[0].SellIn);
            Assert.Equal(9, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesNormalItemsOnSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "normal", SellIn = 0, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-1, Items[0].SellIn);
            Assert.Equal(8, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesNormalItemsAfterSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "normal", SellIn = -5, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-6, Items[0].SellIn);
            Assert.Equal(8, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesNormalItemsWithAQualityOf0()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "normal", SellIn = 5, Quality = 0 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(4, Items[0].SellIn);
            Assert.Equal(0, Items[0].Quality);
        }
        // Brie item
        [Fact]
        public void testUpdatesBrieItemsBeforeSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Aged Brie", SellIn = 5, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(4, Items[0].SellIn);
            Assert.Equal(11, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBrieItemsBeforeSellDateWithMaximumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Aged Brie", SellIn = 5, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(4, Items[0].SellIn);
            Assert.Equal(50, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBrieItemsOnSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Aged Brie", SellIn = 0, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-1, Items[0].SellIn);
            Assert.Equal(12, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBrieItemsOnSellDateWithMaximumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Aged Brie", SellIn = 0, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-1, Items[0].SellIn);
            Assert.Equal(50, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBrieItemsAfterSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Aged Brie", SellIn = -10, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-11, Items[0].SellIn);
            Assert.Equal(12, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBrieItemsAfterSellDateWithMaximumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Aged Brie", SellIn = -10, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-11, Items[0].SellIn);
            Assert.Equal(50, Items[0].Quality);
        }
        // Sulfuras item
        [Fact]
        public void testUpdatesSulfurasItemsBeforeSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Sulfuras, Hand of Ragnaros", SellIn = 5, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(5, Items[0].SellIn);
            Assert.Equal(10, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesSulfurasItemsOnSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Sulfuras, Hand of Ragnaros", SellIn = 0, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(0, Items[0].SellIn);
            Assert.Equal(10, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesSulfurasItemsAfterSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Sulfuras, Hand of Ragnaros", SellIn = -1, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-1, Items[0].SellIn);
            Assert.Equal(10, Items[0].Quality);
        }
        // Backstage Pass
        [Fact]
        public void testUpdatesBackstagePassItemsLongBeforeSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = 11, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(10, Items[0].SellIn);
            Assert.Equal(11, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsCloseToBeforeSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = 10, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(9, Items[0].SellIn);
            Assert.Equal(12, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsCloseToSellDateAtMaximumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = 10, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(9, Items[0].SellIn);
            Assert.Equal(50, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsVeryCloseToSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = 5, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(4, Items[0].SellIn);
            Assert.Equal(13, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsVeryCloseToSellDateAtMaxiumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = 5, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(4, Items[0].SellIn);
            Assert.Equal(50, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsWithOneDayLeftToSellDateAtMaxiumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = 1, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(0, Items[0].SellIn);
            Assert.Equal(50, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsOnSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = 0, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-1, Items[0].SellIn);
            Assert.Equal(0, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsAfterSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Backstage passes to a TAFKAL80ETC concert", SellIn = -1, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-2, Items[0].SellIn);
            Assert.Equal(0, Items[0].Quality);
        }
        // Conjured Mana Cake
        [Fact]
        public void testUpdatesConjuredManaCakeItemsLongBeforeSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = 3, Quality = 6 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(2, Items[0].SellIn);
            Assert.Equal(4, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesConjuredManaCakeItemsCloseToBeforeSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = 10, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(9, Items[0].SellIn);
            Assert.Equal(8, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesConjuredManaCakeItemsCloseToSellDateAtMaximumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = 10, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(9, Items[0].SellIn);
            Assert.Equal(48, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesConjuredManaCakeItemsWithOneDayLeftToSellDateAtMaxiumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = 1, Quality = 50 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(0, Items[0].SellIn);
            Assert.Equal(48, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesConjuredManaCakeItemsOnSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = 0, Quality = 3 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-1, Items[0].SellIn);
            Assert.Equal(0, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesConjuredManaCakeItemsAfterSellDate()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = -1, Quality = 10 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-2, Items[0].SellIn);
            Assert.Equal(6, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsVeryCloseToSellDateAtMiniumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = 5, Quality = 0 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(4, Items[0].SellIn);
            Assert.Equal(0, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsVeryBeforeSellDateAtMiniumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = 10, Quality = 0 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(9, Items[0].SellIn);
            Assert.Equal(0, Items[0].Quality);
        }
        [Fact]
        public void testUpdatesBackstagePassItemsVeryAfterSellDateAtMiniumQuality()
        {
            IList<Item> Items = new List<Item> { new Item { Name = "Conjured Mana Cake", SellIn = -1, Quality = 0 } };
            GildedRose app = new GildedRose(Items);
            app.UpdateQuality();
            Assert.Equal(-2, Items[0].SellIn);
            Assert.Equal(0, Items[0].Quality);
        }
    }
}
